### PR TITLE
[WIP] Detect Null gradient wrt non-sequences in scan

### PR DESCRIPTION
**Pull request by lamblin**
This is a first step towards trying to solve the problem described by Bitton Tenessi at https://groups.google.com/d/msg/theano-users/qgsifuG78KI/rm-oIN_fUbMJ

Still missing:
- [x] tests
- [ ] are other checks (through shared, for instance) needed?

### CODE REVIEW

File: theano/scan_module/scan_op.py [hunk 1]
```diff
@@ -1875,13 +1875,27 @@ def compute_gradient(y, g_y):
...
+        for _idx, y in enumerate(inner_inp_sitsot):
+            x = self.outer_non_seqs(inputs)[_idx]
+            if isinstance(y.type, NullType):
+                # Cannot use dC_dXtm1s.dtype, so we use floatX instead.
+                outer_inp_sitsot.append(
+                    tensor.zeros([grad_steps + 1] +
+                                 [x.shape[i] for i in xrange(x.ndim)],
+                                 dtype=theano.config.floatX))
+                # replace y by a zero tensor of the right shape
+                inner_inp_sitsot[_idx] = tensor.zeros(
```
Comment (abergeron) [inline theano/scan_module/scan_op.py]: This is not safe to do while iterating on the list.  You should write it as a map or something.

File: theano/scan_module/scan_op.py [hunk 2]
```diff
@@ -1875,13 +1875,27 @@ def compute_gradient(y, g_y):
...
-                            self.outer_non_seqs(inputs))]
+        outer_inp_sitsot = []
+        for _idx, y in enumerate(inner_inp_sitsot):
+            x = self.outer_non_seqs(inputs)[_idx]
+            if isinstance(y.type, NullType):
+                # Cannot use dC_dXtm1s.dtype, so we use floatX instead.
+                outer_inp_sitsot.append(
+                    tensor.zeros([grad_steps + 1] +
+                                 [x.shape[i] for i in xrange(x.ndim)],
+                                 dtype=theano.config.floatX))
```
Comment (abergeron) [inline theano/scan_module/scan_op.py]: I will admit that I don't understand why you replace NullTypes with zeros.  I get that it fixes a crash, but is it the right thing to do?

Comment (abergeron) [inline theano/scan_module/scan_op.py]: Oh ok, it's just scan being scan.  It's fine then.

File: theano/scan_module/scan_op.py (hunk 1 above)
Comment (abergeron) [inline theano/scan_module/scan_op.py]: From the python interpreter perspective, yes.  You should not modify lists that you are iterating on.  The documentation states that as long as you don't modify the number of elements it *should* work (like you do here), but I prefer not to rely on that.

The other form you showed gets the length before starting and doesn't touch the list for the iteration counter after so it's safer.

**Comment by carriepl** [comment]
This should cover the "test" part of the PR. I looked at the code and I *think* this is it and additional checks aren't required but... well... it's Scan.

**Comment by lamblin** [comment]
My part was reviewed by @abergeron, and I'm fine with @carriepl's work, so I'm merging.

**Pull request merged**